caffe2/contrib/tensorrt/tensorrt_tranformer.cc

```diff
@@ -0,0 +1,423 @@
+#include "caffe2/contrib/tensorrt/tensorrt_tranformer.h"
+#include "caffe2/contrib/tensorrt/trt_utils.h"
+#include "caffe2/core/logging.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/onnx/onnx_exporter.h"
+#include <onnx2trt.hpp>
+#include <NvInfer.h>
+
+#include <google/protobuf/text_format.h>
+#include <iostream>
+#include <unordered_set>
+
+namespace caffe2 {
+
+namespace {
+
+// TODO(yinghai): Remove the awkward conversion between unordered_map and map
+std::unordered_map<std::string, TensorShape> InferShapes(
+    NetDef* init_net,
+    NetDef* pred_net,
+    const std::unordered_map<std::string, TensorShape>& input_shape_hints) {
+  CaffeMap<std::string, TensorShape> shape_hints_ordered;
+  for (const auto& kv : input_shape_hints) {
+    shape_hints_ordered.emplace(kv.first, kv.second);
+  }
+  std::vector<std::unique_ptr<NetDef>> nets;
+  nets.emplace_back(init_net);
+  nets.emplace_back(pred_net);
+  InferBlobShapesAndTypes(shape_hints_ordered, nets);
+  for (auto& net : nets) {
+    net.release();
+  }
+  std::unordered_map<std::string, TensorShape> shape_hints;
+  for (const auto& kv : shape_hints_ordered) {
+    shape_hints.emplace(kv.first, kv.second);
+  }
+
+  return shape_hints;
+}
+
+std::vector<std::string> FigureInputs(
```

Comment:
This function need better documentation. What is `start` and `end`? Why does it add op outputs to `boundary_inputs`?

Reply:
`start` and `end` defines the continuous chunk of ops that can be readily converted into an TensorRT op. And this function tries to figure out what's the inputs of the to be converted TesnorRT op. What it does it collects the outputs from previous ops, which forms a cut boundary, and they can potential input of the TensorRT op (if referenced).